### ISSUE TITLE
Can't build apollo-ios for watchOS, tvOS using Carthage

### ISSUE THREAD

**Issue by fassko** [apollographql/apollo-ios#386]
- [x] has-reproduction
- [x] blocking

Can't build apollo-ios for macOS, watchOS and tvOS using Carthage.

```
» carthage build apollo-ios --platform macOS,tvOS,watchOS
*** xcodebuild output can be found in /var/folders/jc/ktgtqxj574x18dscvp2vlnfw0000gn/T/carthage-xcodebuild.uRRLh2.log
*** Skipped downloading SQLite.swift.framework binary due to the error:
	"Bad credentials"
*** Skipped downloading Starscream.framework binary due to the error:
	"Bad credentials"
*** Building scheme "SQLite watchOS" in SQLite.xcodeproj
*** Building scheme "SQLite tvOS" in SQLite.xcodeproj
*** Building scheme "SQLite Mac" in SQLite.xcodeproj
*** Building scheme "Starscream" in Starscream.xcodeproj
*** Building scheme "ApolloWebSocket" in Apollo.xcworkspace
*** Building scheme "Apollo" in Apollo.xcworkspace
*** Building scheme "ApolloSQLite" in Apollo.xcworkspace
Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -workspace /Users/kristaps/Documents/qminder/swift-api/Carthage/Checkouts/apollo-ios/Apollo.xcworkspace -scheme ApolloSQLite -configuration Release -derivedDataPath /Users/kristaps/Library/Caches/org.carthage.CarthageKit/DerivedData/10.0_10A255/apollo-ios/0.9.4 -sdk appletvsimulator -destination platform=tvOS\ Simulator,id=BE6DCEB8-5A05-4354-A0E9-C99E0F73EDCC -destination-timeout 3 ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES build (launched in /Users/kristaps/Documents/qminder/swift-api/Carthage/Checkouts/apollo-ios)

This usually indicates that project itself failed to compile. Please check the xcodebuild log for more details: /var/folders/jc/ktgtqxj574x18dscvp2vlnfw0000gn/T/carthage-xcodebuild.uRRLh2.log
```

[carthage-xcodebuild.uRRLh2.log](https://github.com/apollographql/apollo-ios/files/2486759/carthage-xcodebuild.uRRLh2.log)

**Comment by ermik** [comment]
Same.
macOS 10.14
Xcode 10.0; Build version 10A255

**Comment by cannium** [comment]
same issue
macOS 10.14.1
Xcode 10.1 (10B61)

**Comment by mrrhysbutler** [comment]
Is there any update on this?

**Comment by lane-faison** [comment]
Same.
macOS 10.13.6
Xcode 10.1 (10B61)

**Comment by webberwang** [comment]
macOS 10.14.4
Xcode 10.2 (10E125)

**Comment by xlexi** [comment]
Any update to this?

**Comment by fassko** [comment]
@xlexi It does work with Cocoapods, but if you want to integrate this library in your library of framework, then Carthage is the option to use.

**Comment by xlexi** [comment]
Indeed Carthage is preferred, I just provided a temporary solution for others who come across this, since all we're getting from Apollo is tumbleweeds.

**Comment by fawad-khalil** [comment]
Hi, is this issue resolved as Carthage is already in use in our project and we can't use Cocoapods? This is a serious blocker for us :'(

**Comment by designatednerd** [comment]
Sorry about the earlier tumbleweeds and thanks for bringing this to my attention - I'll take a look and see what I can figure out.

**Comment by fawad-khalil** [comment]
Hi @designatednerd , thanks for your consideration. Any updates on this one?

**Comment by designatednerd** [comment]
On my list for this (shortened due to 🇺🇸holiday) week!

**Comment by designatednerd** [comment]
Still looking into what's going on with TV + Watch OS, but if you only need to use either iOS, macOS, or both, you can use `carthage update --platform ios,macos` that currently builds with `0.10.1`.

**Comment by designatednerd** [comment]
Unfortunately, it appears [this is a Carthage issue](https://github.com/Carthage/Carthage/issues/2735).

So the problem is that when things are compiled by Carthage, Carthage is setting the version of a dependent project being compiled incorrectly for both watchOS and tvOS. 

For example, though the `SQLite` dependency is clearly declared as having a 2.2 minimum `watchOS` SDK in its own `xcodeproj`, when using `carthage update --platform watchos` to install it, Carthage compiles the `SQLite` dependency with a 5.2 minimum deployment target, causing the `ApolloSQLite` target to fail because it has a deployment target of 3.0 (which is lower than 5.2). 

I'm reaching out to the Carthage folks to see if they have any suggestions for workarounds, but for now: 

- If you only need iOS or macOS, use the `--platform` argument to specify what platforms you want to build for - those two are building fine at the moment. 
- If you need watchOS and/or tvOS support, you should be able to manually build the frameworks locally and then drag them into your project. This is certainly not ideal but if you really really really can't use CocoaPods, this is probably your best bet. 
- Personally, I think using CocoaPods instead of Carthage is likely the way easier solution until this gets squared on the Carthage side.

**Comment by designatednerd** [comment]
Note: Updated the title of this issue to reflect that macos is now working fine

**Comment by pradipwalghude** [comment]
But issues are not resolved? Provide me the solution asap?

**Comment by pradipwalghude** [comment]
Facing issue on below Xcode version:
macOS 10.14.4
Xcode 10.3
Swift 5.0

**Comment by pradipwalghude** [comment]
> Unfortunately because this is a problem on Carthage's end, there's not a whole lot we can do about it besides [provide these suggested workarounds](https://github.com/apollographql/apollo-ios/issues/386#issuecomment-507247115), and suggest that you leave a comment on the Carthage issue linked in that comment to indicate that this is still a problem you would like to see fixed.

I have uses swift 5 & Xcode 10.3 and Appolo  pod version 0.15.0 . But still facing issue.
Could you please tell where is issue? Have test with latest Xcode ? 

I have installed Apollo by cocoa pods.

**Comment by maximedegreve** [comment]
I'm having the same issue, unfortunately using `--platform ios,macos` isn't working. 

Xcode: 10.3 (10G8)
macOS: 10.14.5
Swift: 5.0

**Comment by maximedegreve** [comment]
Thank you for your response @designatednerd. Unfortunately still getting errors:

**Command:** 
`carthage update --platform macOS --verbose`

**Error:**
`    /bin/sh -c /Users/maximedegreve/Library/Caches/org.carthage.CarthageKit/DerivedData/10.3_10G8/apollo-ios/0.15.0/Build/Intermediates.noindex/ArchiveIntermediates/ApolloWebSocket/IntermediateBuildFilesPath/Apollo.build/Release/Apollo.build/Script-90690D322243442F00FC2E54.sh
env: bash\r: No such file or directory

** ARCHIVE FAILED **


The following build commands failed:
	PhaseScriptExecution Ensure\ no\ build\ settings\ are\ in\ the\ Xcode\ project /Users/maximedegreve/Library/Caches/org.carthage.CarthageKit/DerivedData/10.3_10G8/apollo-ios/0.15.0/Build/Intermediates.noindex/ArchiveIntermediates/ApolloWebSocket/IntermediateBuildFilesPath/Apollo.build/Release/Apollo.build/Script-90690D322243442F00FC2E54.sh
(1 failure)
Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -workspace /Users/maximedegreve/Documents/Projects/MarvelSketch-OSX/Carthage/Checkouts/apollo-ios/Apollo.xcworkspace -scheme ApolloWebSocket -configuration Release -derivedDataPath /Users/maximedegreve/Library/Caches/org.carthage.CarthageKit/DerivedData/10.3_10G8/apollo-ios/0.15.0 ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/79/h65yfsmd0ml5rgqm40tzs6xm0000gn/T/apollo-ios SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /Users/maximedegreve/Documents/Projects/MarvelSketch-OSX/Carthage/Checkouts/apollo-ios)

`

**Comment by maximedegreve** [comment]
Hey @designatednerd any update on this? ☺️

**Comment by designatednerd** [comment]
@maximedegreve Can we move this to a new issue since this is about a `macOS` build rather than watch/tvos? Please let me know there what happens when you open the project from `Carthage/Checkouts/apollo-ios`

**Comment by maximedegreve** [comment]
Yes here you go @designatednerd https://github.com/apollographql/apollo-ios/issues/799

**Comment by kenji21** [comment]
```
$ git clone --recursive https://github.com/apollographql/apollo-ios ; cd apollo-ios
[...]
$ xcodebuild -workspace Apollo.xcworkspace -scheme ApolloSQLite -configuration Release -sdk appletvos
Build settings from command line:
    SDKROOT = appletvos13.0

note: Using new build system
note: Planning build
note: Using build description from disk
Build system information
warning: Multiple targets match implicit dependency for linker flags '-framework SQLite'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'ApolloSQLite' from project 'ApolloSQLite')

[...]

/Users/kenji/Desktop/apollo-ios/Sources/ApolloSQLite/SQLiteSerialization.swift:2:8: error: compiling for tvOS 9.1, but module 'SQLite' has a minimum deployment target of tvOS 13.0: /Users/kenji/Library/Developer/Xcode/DerivedData/Apollo-frvwbgdyhtctxnbmzlggydfgchxp/Build/Products/Release-appletvos/SQLite.framework/Modules/SQLite.swiftmodule/arm64-apple-tvos.swiftmodule
** BUILD FAILED **
```

No change on the 
```
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean

$ git log -1
commit 4f3e862a0fc62790c9d94d17a646d8febb054f6f (HEAD -> master, tag: 0.17.0, origin/master, origin/HEAD)
```

I think fixing this warning could fix the issue : 

```
warning: Multiple targets match implicit dependency for linker flags '-framework SQLite'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'ApolloSQLite' from project 'ApolloSQLite')
```

**Comment by designatednerd** [comment]
It's possible, but I'd be pretty surprised if that's it. The problem seems to be that when Carthage constructs its `xcodebuild` command it's picking the wrong minimum deployment target for the `SQLite` dependency which in turn causes `ApolloSQLite` to fail, not that `xcodebuild` itself isn't working when run directly. 

I'd be very happy if you proved me wrong though, so please feel free to give it a shot!

**Comment by kenji21** [comment]
I was thinking the `-destination` flag added by carthage was causing the issue, so I reduced the `xcodebuild` command to have less options possible

**Comment by kenji21** [comment]
works fine if I inject the deployment target :

```
xcodebuild -workspace Apollo.xcworkspace -scheme ApolloSQLite -configuration Debug -sdk appletvos TVOS_DEPLOYMENT_TARGET=9.1
Build settings from command line:
    SDKROOT = appletvos13.0
    TVOS_DEPLOYMENT_TARGET = 9.1

[...]

** BUILD SUCCEEDED ** [16.569 sec]
```

**Comment by kenji21** [comment]
The issue is not that ApolloSQLite is built for 9.1, but that SQLite.framework is built with min sdk 13.0:

```
otool -l /Users/kenji/Library/Developer/Xcode/DerivedData/Apollo-frvwbgdyhtctxnbmzlggydfgchxp/Build/Products/Debug-appletvos/SQLite.framework/SQLite  | grep LC_BUILD_VERSION -A 5
       cmd LC_BUILD_VERSION
   cmdsize 32
  platform 3
       sdk 13.0
     minos 13.0
    ntools 1
```

**Comment by kenji21** [comment]
I don't understand how SQLite.framework is build, as it is only in referenced in `OTHER_LDFLAGS` and not in dependency:

<img width="933" alt="Screenshot" src="https://user-images.githubusercontent.com/1105089/67110809-deced080-f1d3-11e9-9ff0-e84230076cbf.png">

**Comment by mac-daddie** [comment]
To build today required `xcodebuild -workspace Apollo.xcworkspace -scheme ApolloSQLite -configuration Debug -sdk watchos -WATCHOS_DEPLOYMENT_TARGET=5.3` since otherwise I was getting the error: `SQLiteSerialization.swift:2:8: error: compiling for watchOS 3.0, but module 'SQLite' has a minimum deployment target of watchOS 5.3:`

**Comment by designatednerd** [comment]
That's an interesting workaround - at some point it'd be interesting to dig into this, but honestly my much much stronger recommendation long-term is to use Swift Package Manager instead of Carthage.

**Comment by fassko** [comment]
I agree with @designatednerd I used Carthage to build framework with Apollo SDK, but switched to SwiftPM to do this task. I have no objections about Carthage, I think it's a great tool, but using something made by Apple I think is a safer bet.

**Comment by designatednerd** [comment]
Good news, everyone! Carthage has (finally) released a new version, `0.36.0`, which fixes this issue - I was able to run `carthage build apollo-ios --platform macOS,tvOS,watchOS` without problems. 

At last, we can close this issue out 🎉
